pkgs/development/compilers/flutter/update.py

```diff
@@ -0,0 +1,359 @@
+        """.strip(),
+        "x86_64-darwin": """
+            { fetchzip }: fetchzip {
+              url = "https://storage.googleapis.com/dart-archive/channels/stable/release/%s/sdk/dartsdk-macos-x64-release.zip";
+              sha256 = "";
+            }
+        """.strip(),
+        "aarch64-darwin": """
+            { fetchzip }: fetchzip {
+              url = "https://storage.googleapis.com/dart-archive/channels/stable/release/%s/sdk/dartsdk-macos-arm64-release.zip";
+              sha256 = "";
+            }
+        """.strip(),
+    }
+    result_dict = {}
+    for (platform, code) in code.items():
+        stderr = nix_build_to_fail(code % (dartVersion))
+
+        pattern = re.compile(r"got:\s+(.+?)\n")
+        result_dict[platform] = pattern.findall(stderr)[0]
+
+    return object_to_nix(result_dict)
+
+
+def get_flutter_hash_and_src(flutter_version):
+    code = """
+        {fetchFromGitHub}: fetchFromGitHub {
+          owner = "flutter";
+          repo = "flutter";
+          rev = "%s";
+          hash = "";
+        }
+    """.strip() % (flutter_version)
+
+    stderr = nix_build_to_fail(code)
+    pattern = re.compile(r"got:\s+(.+?)\n")
+    hash = pattern.findall(stderr)[0]
+
+    code = """
+        {fetchFromGitHub}: fetchFromGitHub {
+          owner = "flutter";
+          repo = "flutter";
+          rev = "%s";
+          hash = "%s";
+        }
+    """.strip() % (flutter_version, hash)
+
+    return (hash, nix_build(code))
+
+
+def write_default(nixpkgs_flutter_directory, flutter_version, engine_hash, dart_version, dart_hash, flutter_hash, artifact_hashes):
+    with open(f"{nixpkgs_flutter_directory}/default.nix", "w") as f:
+        f.write(f"""
+{{
+  version = "{flutter_version}";
+  engineVersion = "{engine_hash}";
+  dartVersion = "{dart_version}";
+  dartHash = {dart_hash};
+  flutterHash = "{flutter_hash}";
+  artifactHashes = {artifact_hashes};
+}}
+    """.strip() + "\n")
+
+
+def write_pubspec_lock_json(flutter_compact_version, nixpkgs_flutter_directory, flutter_src):
+    flutter_compact_version = flutter_compact_version.replace('_', '')
+    code = f"{{flutter{flutter_compact_version}}}: flutter{flutter_compact_version}.dart"
+
+    dart_path = f"{nix_build(code)}/bin/dart"
+
+    print(dart_path)
+    print(' '.join([dart_path, 'pub', 'get', '-C',
+          f"{flutter_src}/packages/flutter_tools", '-v']))
+
+    process = subprocess.Popen(
+        [dart_path, 'pub', 'get', '-C',
+            f"{flutter_src}/packages/flutter_tools", '-v'],
+        stderr=subprocess.PIPE,
+        text=None
+    )
+
+    stderr = ""
+    while True:
+        line = process.stderr.readline()
+        if not line:
+            break
+        stderr += line.decode('utf-8')
+
+    process.wait()
+
+    regex = r"FINE: Contents:\n.*Generated by pub(.|\n)*?ERR :"
+    matches = re.search(regex, stderr)
+    pubspec_lock_yaml = '\n'.join(
+        [line[6:] for line in matches.group().strip().splitlines()[1:-1]])
```

Comment:
Can we actually count on this line slice behaviour?

Another approach may be do this inside a fixed output derivation with empty hash on purpose then print a marker in the derivation before and after cat the lockfile and then extract the lockfile from nix logs. The derivation will obvously fail but will output the lockfile in the logs, then just find the markers in the logs and extract what is in between the markers.

Reply:
Advantages of this approach
- Can run on remote builders.
- Harder to misparse, with a marker with exact size of dashes (`echo '----------'` before and after cat pubspec.lock ) it's very unlikely to get it wrong.
- Every user using that will have Nix anyway.
- No stuff will be written in their home folder (`~/.pub`), not even the `result` symlink in this case because the derivation will fail.

Reply:
good idea! done